### PR TITLE
[SYCL][Graph] Implement missing exceptions defined by SYCL-Graphs specification

### PR DESCRIPTION
**Pull request by EwanC**
This PR contains a set of changes that implement throwing an exception on invalid usage, as defined by [sycl_ext_oneapi_graph](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc). Covering the cases missing from the current implementation.

* Error checking for `make_edge` API. 
* Throw exception when explicit add called on a graph recording a queue.
* Throw exception when creating a graph for an unsupported backend.
* Error on invalid buffer behaviour when used with graphs to reflect https://github.com/intel/llvm/pull/10473

## Authors

Co-authored-by: Pablo Reble <pablo.reble@intel.com>
Co-authored-by: Julian Miller <julian.miller@intel.com>
Co-authored-by: Ben Tracy <ben.tracy@codeplay.com>
Co-authored-by: Ewan Crawford <ewan@codeplay.com>
Co-authored-by: Maxime France-Pillois  <maxime.francepillois@codeplay.com>

### CODE REVIEW

File: sycl/source/detail/graph_impl.cpp [hunk 1]
```diff
@@ -12,9 +12,12 @@
 #include <detail/program_manager/program_manager.hpp>
 #include <detail/queue_impl.hpp>
 #include <detail/scheduler/commands.hpp>
+#include <detail/sycl_mem_obj_t.hpp>
 #include <sycl/feature_test.hpp>
 #include <sycl/queue.hpp>
 
+#include <deque>
+
```
Comment (reble) [inline sycl/source/detail/graph_impl.cpp:20]: NIT it's already in the header
```suggestion
```

**Review by reble — APPROVED** [review]

**Review by againull — APPROVED** [review]

File: sycl/test/abi/sycl_symbols_windows.dump [hunk 2]
```diff
@@ -99,8 +99,8 @@
 ??$get_info@Ugpu_slices@device@info@intel@ext@_V1@sycl@@@device_impl@detail@_V1@sycl@@QEBAIXZ
 ??$get_info@Ugpu_subslices_per_slice@device@info@intel@ext@_V1@sycl@@@device@_V1@sycl@@QEBAIXZ
 ??$get_info@Ugpu_subslices_per_slice@device@info@intel@ext@_V1@sycl@@@device_impl@detail@_V1@sycl@@QEBAIXZ
-??$get_info@Ugraph_support@device@info@experimental@oneapi@ext@_V1@sycl@@@device@_V1@sycl@@QEBA?AW4graph_support_level@info@experimental@oneapi@ext@12@XZ
-??$get_info@Ugraph_support@device@info@experimental@oneapi@ext@_V1@sycl@@@device_impl@detail@_V1@sycl@@QEBA?AW4graph_support_level@info@experimental@oneapi@ext@23@XZ
```
Comment (EwanC) [inline sycl/test/abi/sycl_symbols_windows.dump:103]: This change from `experimental::info::graph_support_level` to `experimental::graph_support_level`, I think is an ABI breaking change here as the old symbol is removed.

We're not currently in an ABI break window, so we can revert this part of the change unless you can get an exemption for a symbol not existing in a previous binary release.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
Precommit failure is an infrastructure timeout. Merging.

**Pull request merged**